packages/core/nano/test/util/loggingConsumer.ts

```diff
@@ -0,0 +1,52 @@
+import { IConsumer, IVectorConsumer, IMatrixConsumer, IProducer, IVectorProducer, IMatrixProducer } from "../../src/types";
+import { strict as assert } from "assert";
+
+class NullConsumer<T> implements IConsumer<T>, IVectorConsumer<T>, IMatrixConsumer<T> {
+    public rowsChanged(): void { }
+    public colsChanged(): void { }
+    public cellsReplaced(): void { }
+    public valueChanged(): void {}
+    public itemsChanged(): void {}
+}
+
+/** A generic test consumer that ignores all change notifications. */
+export const nullConsumer: IConsumer<unknown> & IVectorConsumer<unknown> & IMatrixConsumer<unknown> = new NullConsumer<unknown>();
+
+export class LoggingConsumer<T> implements IConsumer<T>, IVectorConsumer<T>, IMatrixConsumer<T> {
+    public readonly log: any[] = [];
+
+    public static setProducerId(producer: any, value: string) { producer.id = value; }
+    private getProducerId(producer: any) { return producer.id || "Error: Missing call to LoggingConsumer.setProducerId(..)"; }
+
+    // #region IConsumer<T>
+    public valueChanged<U extends T, K extends keyof U>(property: K, value: U[K], producer: IProducer<U>): void {
```

Comment:
On a slightly related note: I found that in practice have two generic parameters for `IConsumer` was a pain. I think it was easier to jut have `IProducer<T>` and `IConsumer<T>` and rely on subtyping to allow things to be wider if needed. Something to keep an eye on through iterations.

Reply:
Will do, thx.